File: src/Meridian.FSharp/Domain/SecurityIdentifiers.fs
namespace Meridian.FSharp.Domain

open System

type SecurityId = SecurityId of Guid

[<RequireQualifiedAccess>]
type IdentifierKind =
    | Ticker
    | Isin
    | Cusip
    | Sedol
    | Figi
    | ProviderSymbol of provider:string
    | InternalCode
    /// Legal Entity Identifier (ISO 17442) — 20-char alphanumeric; required for OTC derivatives regulatory reporting.
    | Lei
    /// Refinitiv/LSEG PermID — stable cross-asset persistent identifier.
    | PermId
    /// Bloomberg Global Identifier (BBGID) — stable across corporate actions; distinct from ticker.
    | Bbgid
    /// Wertpapierkennnummer — German/Austrian exchange standard (6 alphanumeric chars).
    | Wkn
    /// Valoren — Swiss SIX exchange security number.
    | Valoren
    /// Meridian-stable ticker that survives corporate actions (Bloomberg PermTicker convention).
    | PermTicker
    /// Reuters Instrument Code — used by Refinitiv Eikon / LSEG feeds.
    | Ric

type Identifier = {
    Kind: IdentifierKind
    Value: string
    IsPrimary: bool
    ValidFrom: DateTimeOffset
    ValidTo: DateTimeOffset option
}

[<RequireQualifiedAccess>]
module SecurityIdentifier =
    let normalizeValue (value: string) =
        if isNull value then String.Empty else value.Trim().ToUpperInvariant()

    let provider identifier =
        match identifier.Kind with
        | IdentifierKind.ProviderSymbol provider -> Some provider
        | _ -> None

    let kindName identifier =
        match identifier.Kind with
        | IdentifierKind.Ticker -> "Ticker"
        | IdentifierKind.Isin -> "Isin"
        | IdentifierKind.Cusip -> "Cusip"
        | IdentifierKind.Sedol -> "Sedol"
        | IdentifierKind.Figi -> "Figi"
        | IdentifierKind.ProviderSymbol _ -> "ProviderSymbol"
        | IdentifierKind.InternalCode -> "InternalCode"
        | IdentifierKind.Lei -> "Lei"
        | IdentifierKind.PermId -> "PermId"
        | IdentifierKind.Bbgid -> "Bbgid"
        | IdentifierKind.Wkn -> "Wkn"
        | IdentifierKind.Valoren -> "Valoren"
        | IdentifierKind.PermTicker -> "PermTicker"
        | IdentifierKind.Ric -> "Ric"

    let isActiveAt asOf identifier =
        identifier.ValidFrom <= asOf
        && identifier.ValidTo |> Option.forall (fun validTo -> validTo > asOf)

    let sameIdentity left right =
        left.Kind = right.Kind
        && String.Equals(normalizeValue left.Value, normalizeValue right.Value, StringComparison.Ordinal)
        && String.Equals(
            left |> provider |> Option.defaultValue String.Empty |> normalizeValue,
            right |> provider |> Option.defaultValue String.Empty |> normalizeValue,
            StringComparison.Ordinal)

    /// Validates an ISIN using the Luhn mod-10 algorithm applied to a numeric-expanded form.
    /// Returns true when the check digit is valid.
    let validateIsin (isin: string) =
        let v = if isNull isin then String.Empty else isin.Trim().ToUpperInvariant()
        if v.Length <> 12 then false
        else
            // Expand letters to digits (A=10 … Z=35) then validate Luhn mod-10
            // Fail fast on any character that is not A-Z or 0-9
            let charValues =
                v
                |> Seq.map (fun c ->
                    if c >= 'A' && c <= 'Z' then
                        let n = int c - int 'A' + 10
                        Some [ n / 10; n % 10 ]
                    elif c >= '0' && c <= '9' then
                        Some [ int c - int '0' ]
                    else None)
                |> Seq.toArray
            if charValues |> Array.exists Option.isNone then false
            else
            let digits =
                charValues
                |> Array.choose id
                |> Array.collect Array.ofList
            let len = digits.Length
            let mutable sum = 0
            let mutable doubleIt = false          // check digit (rightmost) is never doubled in Luhn
            for i = len - 1 downto 0 do
                let mutable d = digits.[i]
                if doubleIt then
                    d <- d * 2
                    if d > 9 then d <- d - 9
                sum <- sum + d
                doubleIt <- not doubleIt
            sum % 10 = 0

    /// Validates a CUSIP (Committee on Uniform Security Identification Procedures) check digit.
    /// Returns true when the check digit is valid.
    let validateCusip (cusip: string) =
        let v = if isNull cusip then String.Empty else cusip.Trim().ToUpperInvariant()
        let tryGetCusipValue c =
            if c >= '0' && c <= '9' then Some (int c - int '0')
            elif c >= 'A' && c <= 'Z' then Some (int c - int 'A' + 10)
            elif c = '*' then Some 36
            elif c = '@' then Some 37
            elif c = '#' then Some 38
            else None

        if v.Length <> 9 then false
        elif v.[8] < '0' || v.[8] > '9' then false
        else
            let digits =
                v
                |> Seq.take 8
                |> Seq.map tryGetCusipValue
                |> Seq.toArray

            if digits |> Array.exists Option.isNone then false
            else
                let sum =
                    digits
                    |> Array.map Option.get
                    |> Array.mapi (fun i d -> if i % 2 = 1 then d * 2 else d)
                    |> Array.sumBy (fun d -> d / 10 + d % 10)

                let check = (10 - (sum % 10)) % 10
                check = (int v.[8] - int '0')
    /// Validates an LEI (Legal Entity Identifier, ISO 17442) using MOD 97-10 (ISO 7064).
    /// Returns true when the check digits are valid.
    let validateLei (lei: string) =
        let v = if isNull lei then String.Empty else lei.Trim().ToUpperInvariant()
        if v.Length <> 20 then false
        else
            // Rearrange: move first 4 chars to end, then expand letters to digits
            let rearranged = v.[4..] + v.[..3]
            let numericParts =
                rearranged
                |> Seq.map (fun c ->
                    if c >= 'A' && c <= 'Z' then Some (string (int c - int 'A' + 10))
                    elif c >= '0' && c <= '9' then Some (string (int c - int '0'))
                    else None)
                |> Seq.toArray
            if numericParts |> Array.contains None then
                false
            else
                let numericStr =
                    numericParts
                    |> Array.choose id
                    |> String.concat ""
                // MOD 97 on large integer via chunked processing
                let mutable remainder = 0
                for chunk in numericStr |> Seq.chunkBySize 9 do
                    let n = System.Int64.Parse(string remainder + System.String(chunk))
                    remainder <- int (n % 97L)
                remainder = 1

File: src/Meridian.Contracts/SecurityMaster/SecurityIdentifiers.cs
using System.Text.Json.Serialization;

namespace Meridian.Contracts.SecurityMaster;

[JsonConverter(typeof(JsonStringEnumConverter<SecurityIdentifierKind>))]
public enum SecurityIdentifierKind
{
    Ticker,
    Isin,
    Cusip,
    Sedol,
    Figi,
    /// <summary>OCC/OSI listed-option contract symbol such as AAPL240621C00150000.</summary>
    OccOptionSymbol,
    ProviderSymbol,
    InternalCode,
    /// <summary>Legal Entity Identifier (ISO 17442) — 20-char alphanumeric; required for OTC derivatives regulatory reporting.</summary>
    Lei,
    /// <summary>Refinitiv/LSEG PermID — stable cross-asset persistent identifier.</summary>
    PermId,
    /// <summary>Bloomberg Global Identifier (BBGID) — stable across corporate actions; distinct from ticker.</summary>
    Bbgid,
    /// <summary>Wertpapierkennnummer — German/Austrian exchange standard (6 alphanumeric chars).</summary>
    Wkn,
    /// <summary>Valoren — Swiss SIX exchange security number.</summary>
    Valoren,
    /// <summary>Meridian-stable ticker that survives corporate actions (Bloomberg PermTicker convention).</summary>
    PermTicker,
    /// <summary>Reuters Instrument Code — used by Refinitiv Eikon / LSEG feeds.</summary>
    Ric,
    /// <summary>SEC Central Index Key — identifies an EDGAR filer or issuer, not a standalone tradable security.</summary>
    Cik,
    /// <summary>
    /// Read-tolerance member: a kind written by a newer node that this node does not recognize.
    /// Rows carrying it stay readable (mixed-version rollout) but cannot be written back as-is —
    /// the command mapping rejects <see cref="Unknown"/> so unrecognized kinds are never silently
    /// re-persisted.
    /// </summary>
    Unknown
}

[JsonConverter(typeof(JsonStringEnumConverter<SecurityAliasScope>))]
public enum SecurityAliasScope
{
    Operations,
    Collector,
    Execution,
    Migration,
    /// <summary>Read-tolerance member for scopes written by newer nodes; see <see cref="SecurityIdentifierKind.Unknown"/>.</summary>
    Unknown
}

public sealed record SecurityIdentifierDto(
    SecurityIdentifierKind Kind,
    string Value,
    bool IsPrimary,
    DateTimeOffset ValidFrom,
    DateTimeOffset? ValidTo = null,
    string? Provider = null,
    string? NormalizedValue = null,
    string? NormalizedProvider = null);

public sealed record SecurityAliasDto(
    Guid AliasId,
    Guid SecurityId,
    string AliasKind,
    string AliasValue,
    string? Provider,
    SecurityAliasScope Scope,
    string? Reason,
    string CreatedBy,
    DateTimeOffset CreatedAt,
    DateTimeOffset ValidFrom,
    DateTimeOffset? ValidTo,
    bool IsEnabled);

/// <summary>
/// Raised when a request would mutate an existing alias row without an append-only alias revision.
/// The current projection table is consumed by recorded-as-of reads, so an in-place correction
/// would rewrite history and must be reported as a state conflict.
/// </summary>
public sealed class SecurityAliasHistoryConflictException : InvalidOperationException
{
    public SecurityAliasHistoryConflictException(Guid aliasId)
        : base(
            $"Alias '{aliasId}' already exists with different recorded facts. " +
            "In-place replacement or retirement is unavailable until append-only alias revisions are implemented.")
    {
        AliasId = aliasId;
    }

    public Guid AliasId { get; }
}
